### PR TITLE
iox-#1196 Fix clang-tidy warning in storable_function

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1196

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.hpp [hunk 1]
```diff
@@ -65,9 +65,17 @@ class storable_function<StorageType, signature<ReturnType, Args...>>
               typename = typename std::enable_if<std::is_class<Functor>::value
                                                      && is_invocable_r<ReturnType, Functor, Args...>::value,
                                                  void>::type>
+    /// @NOLINTJUSTIFICATION the storable function should implicitly behave like any generic constructor, adding
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.hpp:68]: Same as mentioned [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1495#discussion_r920005056), the `@` can be removed.

```suggestion
    /// NOLINTJUSTIFICATION the storable function should implicitly behave like any generic constructor, adding
```

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp [hunk 2]
```diff
@@ -53,13 +53,13 @@ class Counter
         ++numCreated;
     }
 
-    Counter(const Counter&)
+    Counter(const Counter& rhs IOX_MAYBE_UNUSED)
```
Comment (mossmaurice) [inline iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:56]: Nitpick: Could also be
```suggestion
    Counter(const Counter&)
```

**Review by mossmaurice — DISMISSED** [review]
LGTM, thanks!

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp [hunk 3]
```diff
@@ -53,13 +53,13 @@ class Counter
         ++numCreated;
```
Comment (mossmaurice) [inline iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:53]: @elfenpiff Could you please adapt the copyright header?

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:56]: No then clang-tidy prints out a warning!

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl [hunk 4]
```diff
@@ -37,6 +37,9 @@ storable_function<S, signature<ReturnType, Args...>>::storable_function(ReturnTy
     if (function)
     {
         m_invoker = invokeFreeFunction;
+        /// @NOLINTJUSTIFICATION we use type erasure in combination with compile time template arguments to restore
+        ///                      to correct type whenever the callable is used
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:41]:
```suggestion
        ///                      the correct type whenever the callable is used
```

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp (hunk 2 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:56]: Yes but I think this is due to `hicpp-named-parameter` which I am in favor of disabling (have it disabled in https://github.com/eclipse-iceoryx/iceoryx/pull/1475/files#diff-0534891fbc8b89b4c521057e7e8dff70d3cbeb822cc9991a3abad77848c289f8). It does more harm than good and is too restrictive. The compiler forces us to have names where they are required anyway; it is impossible to circumvent it for variables that are really used, and for others it should not be mandatory (due to cases like this).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl (hunk 4 above)
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:41]:
```suggestion
        ///                      the correct type whenever the callable is used
```

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl [hunk 5]
```diff
@@ -284,6 +293,9 @@ ReturnType storable_function<S, signature<ReturnType, Args...>>::invoke(void* ca
 template <typename S, typename ReturnType, typename... Args>
 ReturnType storable_function<S, signature<ReturnType, Args...>>::invokeFreeFunction(void* callable, Args&&... args)
 {
+    /// @NOLINTJUSTIFICATION we use type erasure in combination with compile time template arguments to restore
+    ///                      to correct type whenever the callable is used
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/storable_function.inl:297]:
```suggestion
    ///                      the correct type whenever the callable is used
```

File: iceoryx_hoofs/test/moduletests/test_cxx_function.cpp [hunk 6]
```diff
@@ -69,14 +69,22 @@ class Counter
         ++numDestroyed;
     }
 
-    Counter& operator=(const Counter&)
+    Counter& operator=(const Counter& rhs)
     {
-        ++numCopied;
+        if (this != &rhs)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/test/moduletests/test_cxx_function.cpp:74]: I am not sure it was initially meant to count the invocations of this Ctor or the actual copies. I think the latter. It should it not be used in self copy/move scenarios in the tests, so either is fine.

**Review by MatthiasKillat — DISMISSED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]

**Pull request merged**